pennylane/hf/spin.py

```diff
@@ -0,0 +1,328 @@
+    r"""Computes the total spin projection observable :math:`\hat{S}_z`.
+
+    The total spin projection operator :math:`\hat{S}_z` is given by
+
+    .. math::
+
+        \hat{S}_z = \sum_{\alpha, \beta} \langle \alpha \vert \hat{s}_z \vert \beta \rangle
+        ~ \hat{c}_\alpha^\dagger \hat{c}_\beta, ~~ \langle \alpha \vert \hat{s}_z
+        \vert \beta \rangle = s_{z_\alpha} \delta_{\alpha,\beta},
+
+    where :math:`s_{z_\alpha} = \pm 1/2` is the spin-projection of the single-particle state
+    :math:`\vert \alpha \rangle`. The operators :math:`\hat{c}^\dagger` and :math:`\hat{c}`
+    are the particle creation and annihilation operators, respectively.
+
+    Args:
+        orbitals (str): Number of *spin* orbitals. If an active space is defined, this is
+            the number of active spin-orbitals.
+
+    Returns:
+        pennylane.Hamiltonian: the total spin projection observable :math:`\hat{S}_z`
+
+    **Example**
+
+    >>> orbitals = 4
+    >>> print(spin_z(orbitals))
+    (-0.25) [Z0]
+    + (0.25) [Z1]
+    + (-0.25) [Z2]
+    + (0.25) [Z3]
+    """
+
+    if orbitals <= 0:
+        raise ValueError(f"'orbitals' must be greater than 0; got for 'orbitals' {orbitals}")
+
+    r = np.arange(orbitals)
+    sz_orb = np.where(r % 2 == 0, 0.5, -0.5)
+    table = np.vstack([r, r, sz_orb]).T
+
+    sz_coeff = np.array([])
+    sz_op = []
+
+    for i in table:
+        sz_coeff = np.concatenate((sz_coeff, np.array([i[2]])))
+        sz_op.append([int(i[0]), int(i[1])])
+
+    return qubit_operator((sz_coeff, sz_op))
+
+
+def particle_number(orbitals):
+    r"""Computes the particle number observable :math:`\hat{N}=\sum_\alpha \hat{n}_\alpha`
+    in the Pauli basis.
+
+    The particle number operator is given by
+
+    .. math::
+
+        \hat{N} = \sum_\alpha \hat{c}_\alpha^\dagger \hat{c}_\alpha,
+
+    where the index :math:`\alpha` runs over the basis of single-particle states
+    :math:`\vert \alpha \rangle`, and the operators :math:`\hat{c}^\dagger` and :math:`\hat{c}` are
+    the particle creation and annihilation operators, respectively.
+
+    Args:
+        orbitals (int): Number of *spin* orbitals. If an active space is defined, this is
+            the number of active spin-orbitals.
+
+    Returns:
+        pennylane.Hamiltonian: the particle number observable
+
+    **Example**
+
+    >>> orbitals = 4
+    >>> print(particle_number(orbitals))
+    (2.0) [I0]
+    + (-0.5) [Z0]
+    + (-0.5) [Z1]
+    + (-0.5) [Z2]
+    + (-0.5) [Z3]
+    """
+
+    if orbitals <= 0:
+        raise ValueError(f"'orbitals' must be greater than 0; got for 'orbitals' {orbitals}")
+
+    r = np.arange(orbitals)
+    table = np.vstack([r, r, np.ones([orbitals])]).T
+
+    coeffs = np.array([])
+    ops = []
+
+    for i in table:
+        coeffs = np.concatenate((coeffs, np.array([i[2]])))
+        ops.append([int(i[0]), int(i[1])])
+
+    return qubit_operator((coeffs, ops))
+
+
+def qubit_operator(o_ferm, cutoff=1.0e-12):
```

Comment:
This function is added in a previous [PR](https://github.com/PennyLaneAI/pennylane/pull/2173). Copying it here to pass the tests.